File: İlker Avcı/03.Hafta/quicksort.py
#!/usr/bin/env python3

def partition(arr: list, low: int, high: int) -> int:
    # pivot will be the last element
    pivot = arr[high]
    i = low - 1
    j = low
    while j <= high - 1:
        if arr[j] < pivot:
            i += 1

            #swap arr[i] and arr[j] if arr[j] is smaller than the pivot
            arr[i], arr[j] = arr[j], arr[i]

            j += 1

    #swap arr[i+1] and arr[high] (which is pivot)
    arr[i + 1], arr[high] = arr[high], arr[i + 1]

    return i + 1


def quicksort(arr: list, low: int, high: int):
    # Nothing to sort if low is greater than or equal to high
    if low >= high:
        return

    part_index = partition(arr, low, high)

    quicksort(arr, low, part_index - 1)  # Left side of the partitioning index
    quicksort(arr, part_index + 1, high)  # Right side of the partitioning index

File: Numan Kafadar/week3/quicksort.cpp
#include <iostream>
using namespace std;

int partition(int arr[],int left,int right){
    
    int begin=left-1;
    int end=arr[right];
    for(int i=left;i<=right-1;i++){
        if(arr[i]<end){
            begin++;
            swap(arr[begin],arr[i]);
        }
    }
    swap(arr[begin+1] ,arr[right] );
    return(begin+1);
}


void quicksort(int arr[],int left,int right){
    int pivot;
    if(left<right){
        pivot=partition(arr,left,right);
        quicksort(arr,left,pivot-1);
        quicksort(arr,pivot+1,right);
    }

}


int main()
{
    int arr[] = {66,10, 7,21, 8, 111, 86,19, 9, 1, 5,6,12,17,28};
    int n = sizeof(arr) / sizeof(arr[0]);
    quicksort(arr, 0, n-1);
    cout << "Sorted array: ";
    for(int i=0;i<n;i++){
        cout<<arr[i]<<" ";
    }
    cout<<endl;
    return 0;
}